earth_enterprise/src/fusion/portableglobe/shared/bounds_tracker.cpp

```diff
@@ -0,0 +1,117 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+// Methods for keeping track of quadtree bounds and writing them to a JSON file.
+
+#include "fusion/portableglobe/shared/bounds_tracker.h"
+#include <fstream>  // NOLINT(readability/streams)
+#include <iostream>  // NOLINT(readability/streams)
+#include <string>
+#include "common/khFileUtils.h"
+#include "common/khSimpleException.h"
+#include "fusion/portableglobe/quadtree/qtutils.h"
+#include "fusion/portableglobe/shared/packetbundle.h"
+
+namespace fusion_portableglobe {
+
+BoundsTracker::BoundsTracker() {}
+
+void BoundsTracker::update(uint16_t channel_id, PacketType type, const std::string& qtpath) {
+  std::string real_path = "0" + qtpath;
+
+  uint32_t column, row, zoom;
+  ConvertFromQtNode(real_path, &column, &row, &zoom);
+
+  uint32_t level = qtpath.size();
+
+  if (channels.count(channel_id) == 0) {
+    channels[channel_id] = {channel_id, type, row, column, level};
+
+  } else {
+    channel_info& channel = channels[channel_id];
+
+    // BoundsTracker only tracks bounds at the maximum zoom level (max_level)
+    if (level < channel.min_level) {
+      channel.min_level = level;
+
+      // If this node is at a new max_level then set max_level and also update the bounds
+      // to be the current row and column.
+    } else if (level > channel.max_level) {
+      channel.max_level = level;
+      channel.left = channel.right = column;
+      channel.top = channel.bottom = row;
+
+      // A new node at max_level, so expand the bounding box if necessary.
+    } else if (level == channel.max_level) {
+      channel.left = std::min(column, channel.left);
+      channel.right = std::max(column, channel.right);
+      channel.top = std::min(row, channel.top);
+      channel.bottom = std::max(row, channel.bottom);
+    }
+  }
+}
+
+void BoundsTracker::write_json_file(const std::string& filename) const {
+  khEnsureParentDir(filename);
+
+  static std::map<PacketType, std::string> channel_type_strings =
+    {{ kDbRootPacket,  "DbRoot"},
+     { kDbRoot2Packet, "DbRoot2"},
+     { kQtpPacket,     "Qtp"},
+     { kQtp2Packet,    "Qtp2"},
+     { kImagePacket,   "Image"},
+     { kTerrainPacket, "Terrain"},
+     { kVectorPacket,  "Vector"}};
+    
+  std::ofstream fout(filename.c_str());
+
+  // Find the last entry so the last comma can be skipped
+  auto last_entry = channels.end();
+  --last_entry;
+
+  fout << "[\n";
+  for (auto iter =  channels.begin(); iter != channels.end(); ++iter) {
```

Comment:
Actually, this needs to use the begin()/end() style loop so that it can check if the iterator is pointing at the last element when it decides to print a comma or not.